BlockAssembler: return selected packages virtual size and fee

ismaelsadeeq opened the pull request:
This PR enables `BlockAssembler` to add all selected packages' fee and virtual size to a vector, and then return the vector as a member of `CBlockTemplate` struct.

This PR is the first step in the https://github.com/bitcoin/bitcoin/issues/30392 project.

The packages' vsize and fee are used in #30157 to select a percentile fee rate of the top block in the mempool.

ismaelsadeeq commented on src/policy/feerate.cpp:
This causes a fuzz [failure](https://cirrus-ci.com/task/4852900168466432) during the conversion of the size to feerate per KB because `CAmount` is `int64_t`.

So I reverted back and `static_cast` the `packageSize` to be `uint32_t`, it is safe to do so because we are certain that `nSizeWithAncestors` will always be positive in `BlockAssembler::addPackageTxs`.

---
It's already done https://github.com/bitcoin/bitcoin/blob/bd5d1688b4311e21c0e0ff89a3ae02ef7d0543b8/src/node/miner.cpp#L365

ismaelsadeeq replied:
Thanks Yeah. 
---
unrelated but I wonder why we are using this big data type to represent size, not only while working in this PR, there are size datype inconsistencies in several places which sometimes cause overflows during calculation.

Is it safe to instead represent a type alias for size just like it was done for `CAmount`  in `consensus/amount.h`.
And prevent generating large values for size in the fuzz test?

Apart from the fuzz test size can never be greater than more than `4,000,000`?
https://github.com/bitcoin/bitcoin/blob/bd5d1688b4311e21c0e0ff89a3ae02ef7d0543b8/src/consensus/consensus.h#L12-L15

ismaelsadeeq replied:
> The fuzz failure is because you also changed `MultiplicationOverflow` to `uint64_t`, which is wrong.

Yeah, I did that because `MultiplicationOverflow` requires that the type of both parameters be the same.

ismaelsadeeq replied:
thanks the commit is gone now.

ismaelsadeeq commented on src/node/miner.h:
```diff
@@ -38,6 +40,7 @@ struct CBlockTemplate
     std::vector<CAmount> vTxFees;
     std::vector<int64_t> vTxSigOpsCost;
     std::vector<unsigned char> vchCoinbaseCommitment;
+    std::vector<std::tuple<CFeeRate, uint32_t>> vFeeratePerSize;
```
@glozow 
> probably `FeeFrac` would be better? `CFeeRate` loses precision and can be inferred from the fee and size

We will be using the package size to calculate the percentile package and select its fee rate as an estimate to return to the user.

We then have to convert the fee and size to fee rate either by `CFeeRate(const CAmount& nFeePaid, uint32_t num_bytes).GetFeePerK` or introduce and use a similar `GetFeeRate` method for the `FeeFrac` struct.

So, I think using `CFeeRate` is okay for getting `MEMPOOL_FORECAST` estimate.

---
I think the place we could benefit from changing this to `FeeFrac` is when comparing forecasters’ estimates.

But this can’t be done now because the `BLOCK_POLICY_ESTIMATOR` forecaster performs its operations in `nSatoshiPerK` and returns a `CFeeRate `value as an estimate.

It will require a refactor to change `BLOCK_POLICY_ESTIMATOR` to use and return `FeeFrac` as an estimate. 

But the forecasters estimates are values greater than 1 s/vb. 
Given this, I think using `CFeeRate` is okay and will avoid the conversions, or adding duplicate `GetFeePerK` method for `FeeFrac` and refactoring `CBlockPolicyEstimator` to use `FeeFrac`?

glozow replied:
Nonblocking, I can understand it's a larger change to use it.

ismaelsadeeq replied:
I've updated the code to use `FeeFrac`, and it simplified things – no need for tuples from the initial approach, just a vector of `FeeFrac` values which achieved same aim.

Smaller diff and no precision loss 💯 
I will update all the dependent commits to follow suite.

rkrux commented on src/test/miner_tests.cpp:
```diff
@@ -122,26 +144,49 @@ void MinerTestingSetup::TestPackageSelection(const CScript& scriptPubKey, const
+    const auto packageSize = lowFeeTx.GetTxSize() + highFeeChildTx.GetTxSize();
+
+    CFeeRate packageRate(packageFee, packageSize);
+    auto packageFeeRateFound = findFeeRate(blockFeeAndVsizes, packageRate);
+
+    CFeeRate mediumRate(mediumFeeTx.GetFee(), mediumFeeTx.GetTxSize());
+    auto mediumTxFeeRateFound = findFeeRate(blockFeeAndVsizes, mediumRate);
```
IMO, depending upon finding **only** the feeRate in the vector of tuples gets rid of some level of robustness in the verification here because it ignores the possibility of duplicate feeRates, and makes the tester/author wary of writing the test in a way that ensures there are no duplicates. 
How about **also** checking for the packageSize as well?

ismaelsadeeq replied:
That makes sense. I removed `findFeeRate` and explicitly checked the vector for the packages and verify the fee and size were as expected. I also took the opportunity to change the variables to be initialized uniformly.

rkrux replied:
Thank you for accepting the suggestion, the diff now has become smaller and to the point.

rkrux reviewed (approved):
tACK [f581657](https://github.com/bitcoin/bitcoin/commit/f581657b3a1acd0fc57dc9461d3a7e0835e17a24)

`make, make check` are successful. Left a suggestion.
Thanks for keeping the PR short and focussed.

rkrux reviewed (approved):
reACK [8a830f1](https://github.com/bitcoin/bitcoin/commit/8a830f15cf4923f57abf0bdf30b98b04b47c6e7b)

Sjors commented:
Just a comment about how this interacts with the Mining interface changes I'm working on. 

This PR adds a new field to CBlockTemplate. The Mining interface contains BlockTemplate:

https://github.com/bitcoin/bitcoin/blob/018e5fcc462caebb25cf5d3eb6a19dc2787466c8/src/interfaces/mining.h#L31-L40

It has a method to return the `CBlockHeader`. I wonder if adding a field to CBlockTemplate would cause a breaking change for clients connecting via IPC. cc @ryanofsky?

In any case it won't matter as long as this PR is merged before v29, since we don't ship the interface until then, so there's clients to break.

ismaelsadeeq commented:
> So there are clients that this could break.

@Sjors, could you clarify your coment?
Which clients would this affect?

Sjors commented:
@ismaelsadeeq are you referring to my statement at the end "so there's clients to break"?

Once we release the Mining interface, and multiprocess, people can build clients. If we then make changes in the next release, we could break those clients. But that's not an issue yet.

glozow reviewed (commented):
ACK 5171f16e6f96030fcc91b30bb196a2790e37848c

ryanofsky commented on src/node/miner.h:
```diff
@@ -148,6 +150,7 @@ class BlockAssembler
     uint64_t nBlockSigOpsCost;
     CAmount nFees;
     std::unordered_set<Txid, SaltedTxidHasher> inBlock;
+    std::vector<FeeFrac> feerateHistogram;
```
In commit "miner: `CreateNewBlock` will now track fee and virtual size of selected packages" (5171f16e6f96030fcc91b30bb196a2790e37848c)

It seems like it would be simpler and less fragile if `BlockAssembler::addPackageTxs` just populated `pblocktemplate->FeerateHistogram` directly, instead of filling this new `BlockAssembler::feerateHistogram` variable, which is effectively a temporary variable, and then moved from `BlockAssembler::feerateHistogram` to `pblocktemplate->FeerateHistogram`. It would be nice to be able to drop this new variable, and drop the new code moving the histogram from one variable to another.

ismaelsadeeq replied:
Good Suggestion.
Done!

ryanofsky reviewed (approved):
Code review ACK 5171f16e6f96030fcc91b30bb196a2790e37848c

> fwiw added to milestone and this seems like a fairly simple change. @Sjors @ryanofsky would you mind taking another look? I don't think I have full context on Mining interface.

There's actually no risk of this change breaking the mining interface because the mining interface isn't directly exposing the CBlockTemplate struct. Also even if the CBlockTemplate struct were directly exposed it wouldn't necessarily be a breaking change because old clients or servers encountering the new field would just ignore it, and new clients or servers expecting the new field could detect that the field is missing and handle that case appropriately.

ismaelsadeeq commented:
Thank you for your review, @glozow and @ryanofsky.  
I've force-pushed to rebase and address https://github.com/bitcoin/bitcoin/pull/30391#discussion_r1889098891

Sjors commented:
I'm fine with adding `vFeerateHistogram` to `CBlockTemplate`. I haven't looked into https://github.com/bitcoin/bitcoin/pull/30157 so I don't have a strong take on the rest of this PR. Seems pretty straightforward though.

ryanofsky reviewed (approved):
Code review ACK 7de205d2c12a0e5c0dee68728b088b022f63e84d. Seems very simple now, just a two line code change + tests

glozow reviewed (commented):
ACK 7de205d2c12a0e5c0dee68728b088b022f63e84d

rkrux commented on src/node/miner.cpp:
```diff
@@ -421,6 +421,7 @@ void BlockAssembler::addPackageTxs(int& nPackagesSelected, int& nDescendantsUpda
         }
 
         ++nPackagesSelected;
+        pblocktemplate->vFeerateHistogram.emplace_back(packageFees, static_cast<int32_t>(packageSize));
```
https://en.cppreference.com/w/cpp/container/vector/emplace_back#:~:text=It%20demonstrates%20how%20emplace_back%20forwards%20parameters%20to%20the%20President%20constructor

This is handy!

rkrux commented on src/node/miner.h:
```diff
@@ -39,6 +40,7 @@ struct CBlockTemplate
     std::vector<CAmount> vTxFees;
     std::vector<int64_t> vTxSigOpsCost;
     std::vector<unsigned char> vchCoinbaseCommitment;
+    std::vector<FeeFrac> vFeerateHistogram;
```
Getting thrown off a bit by the name here. Is it really a histogram at this point or just a vector of fee rates/fracs?
Going by this def: https://en.wikipedia.org/wiki/Histogram

Sorry for not raising this earlier.

rkrux replied:
Might be helpful to add a comment here telling that the values in this vector are sorted. That's what I assumed while going through the `CalculatePercentiles` function here: https://github.com/bitcoin/bitcoin/pull/30157/files#diff-34fc4771c305399b6f7634acd8c5fd2d0b89115f998397d41676180d742c28daR10

ismaelsadeeq replied:
> Is it really a histogram at this point or just a vector of fee rates/fracs?  

It is a vector of package fee frac's, which can be used to build a histogram.   

> Might be helpful to add a comment here telling that the values in this vector are sorted. That's what I assumed while going through the `CalculatePercentiles` function here.  

Yes, it is sorted by the order in which the packages are selected in the block template. I believe this is obvious because we are emplacing the fee frac's after adding the selected package.
However, adding a comment could be useful for clarity.  

I will update if there is a need to retouch.

rkrux replied:
> It is a vector of package fee frac's, which can be used to build a histogram.

Building a histogram seems to be one of the use cases of this field. I don't feel it should be called a histogram only because of this. Is there any other reason as well?
I hope I'm not missing any context here.

ismaelsadeeq replied:
I think you are, I'm not introducing this term in this PR
You can see https://github.com/bitcoin/bitcoin/pull/21422 for context

ismaelsadeeq replied:
I've seen the vectors above `vFeerateHistogram`, are also simple `vector` types containing `integers`, `unsigned char`s, or `CAmount`s.

However, we do not name these vectors `vCAmounts`, `vInts`, or `vUnsignedChars`. Instead, they are named based on some context, such as `vTxFees`, `vTxSigOpsCost`, and `vchCoinbaseCommitment`.
For this reason, I believe naming this vector `vFeeFrac` does not add much clarity because thats obvious. 
Instead, I consider `vFeerateHistogram` to be the closest meaningful name, based on the context of #21422.

Consider this block template vector of package of fee rates: [20/5, 20/5, 20/5, 20/6, 20/6, 15/6].
after sorting by increasing feerate this implicitly represent a histogram-like structure:
- The x-axis could represent distinct fee rate packages.  
- The y-axis could represent their frequencies.

Same as with  @glozow comment in https://github.com/bitcoin/bitcoin/pull/30391#discussion_r1904219603.

This vector is also implicitly a fee rate diagram. As we can use it to construct a feerate diagram after sorting by increasing feerate.

- The points in the x axis would be the accumulated size of the packages from i to n.
- The points in the y axis would be the fees of each package from i to n.

@rkrux If a more explicit name is desired, `vPackagesFeerates` would be the most accurate IMO.

We can continue the [discussion](https://bikeshed.com) , but since this PR already has two A CKs I was hesitating to update, I will proceed to update the variable name and add a comment if a rebase or blocking suggestion arises.

Let me know if this a blocking comment, I am happy to update.

rkrux replied:
> However, we do not name these vectors vCAmounts, vInts, or vUnsignedChars. Instead, they are named based on some context, such as vTxFees, vTxSigOpsCost, and vchCoinbaseCommitment.
For this reason, I believe naming this vector vFeeFrac does not add much clarity because thats obvious.
Instead, I consider vFeerateHistogram to be the closest meaningful name, based on the context of https://github.com/bitcoin/bitcoin/pull/21422.

I agree with naming variables based on context, this is quite a common practice. I have raised this point because as a reviewer I found it confusing to see a field that's called a histogram but is not, and therefore I don't think it is the closest meaningful name. Either `feeRateDiagram` or `packageFeeRates` sound good to me.  

> We can continue the [discussion](https://bikeshed.com/) , but since this PR already has two A CKs I was hesitating to update, I will proceed to update the variable name and add a comment if a rebase or blocking suggestion arises.

I understand that naming variables can be in the bikeshed territory and as an author it's difficult to incorporate changes after getting multiple ACKs. That's why I didn't request changes and instead started a discussion because to me this name is confusing and was even more so to read it in the test, and thus I could not ACK the PR.

ismaelsadeeq replied:
Fixed now.
Thanks for the review.

rkrux reviewed (commented):
Nit: In the PR description, a draft PR is mentioned as the project. Can add this link for project issue: https://github.com/bitcoin/bitcoin/issues/30392

ismaelsadeeq commented:
> Nit: In the PR description, a draft PR is mentioned as the project. Can add this link for project issue: #30392

Added, and addressed your comments!

rkrux commented on src/test/miner_tests.cpp:
```diff
@@ -121,28 +122,43 @@ void MinerTestingSetup::TestPackageSelection(const CScript& scriptPubKey, const
+    const auto packageFee{lowFeeTx.GetFee() + highFeeChildTx.GetFee()};
+    const auto packageSize{lowFeeTx.GetTxSize() + highFeeChildTx.GetTxSize()};
+    FeeFrac packageFeeFrac{packageFee, packageSize};
+    BOOST_CHECK(blockFeerateHistogram[0] == packageFeeFrac);
```
Related to [this comment](https://github.com/bitcoin/bitcoin/pull/30391/files#r1904134687), I find the language here confusing as well. A fee rate histogram element, which I would expect to be a feeRate interval/group, is being compared to a feeFrac element, which is just a pair of fee and size.

glozow replied:
Ah, @ismaelsadeeq did you mean to call this a feerate *diagram* instead?

@rkrux We've previously represented feerate diagrams as a vector of `FeeFrac`s / chunks
- https://delvingbitcoin.org/t/mempool-incentive-compatibility/553#feerate-diagrams-as-a-way-to-resolve-this-question-3
- #29242

achow101 commented on src/node/miner.h:
```diff
@@ -39,6 +40,7 @@ struct CBlockTemplate
     std::vector<CAmount> vTxFees;
     std::vector<int64_t> vTxSigOpsCost;
     std::vector<unsigned char> vchCoinbaseCommitment;
+    std::vector<FeeFrac> vFeerateHistogram;
```
nit: The style guide says to use the new style for new code, regardless of the style of the surrounding code. So this should use `snake_case` with `m_` prefix: `m_feerate_histogram`.

ismaelsadeeq replied:
fixed

achow101 commented on src/test/miner_tests.cpp:
```diff
@@ -121,28 +122,43 @@ void MinerTestingSetup::TestPackageSelection(const CScript& scriptPubKey, const
     // This tx has a low fee: 1000 satoshis
     Txid hashParentTx = tx.GetHash(); // save this txid for later use
-    AddToMempool(tx_mempool, entry.Fee(1000).Time(Now<NodeSeconds>()).SpendsCoinbase(true).FromTx(tx));
+    const auto lowFeeTx{entry.Fee(1000).Time(Now<NodeSeconds>()).SpendsCoinbase(true).FromTx(tx)};
```
nit: This should be named similarly to the txid to avoid confusion that this might be a different tx altogether. I suggest either calling this `parent_tx` or renaming `hashParentTx` to `hash_low_fee_tx`.

Also `snake_case` here and elsewhere in this test.

ismaelsadeeq replied:
Thanks this is fixed, I'll use `snake_case` henceforth.

ismaelsadeeq committed:
miner: add package feerate vector to CBlockTemplate

- The package feerates are ordered by the sequence in which
  packages are selected for inclusion in the block template.

- The commit also tests this new behaviour.

Co-authored-by: willcl-ark <will@256k1.dev>

rkrux reviewed (approved):
tACK 7c123c08ddce87ae55abfa83bea4a691bc0bd5e7

Thank you for incorporating the changes.

ryanofsky reviewed (approved):
Code review ACK 7c123c08ddce87ae55abfa83bea4a691bc0bd5e7. Changes since last review are rebasing due to a test conflict, giving the new field a better name and description, resolving the test conflict, and renaming a lot of test variables. The actual code change is still one-line change.

glozow reviewed (commented):
reACK 7c123c08ddce87ae55abfa83bea4a691bc0bd5e7
